etl/command.py

```diff
@@ -270,22 +287,100 @@ def run_dag(
+            time_taken = timed_run(lambda: step.run())
+            click.echo(f"{click.style('OK', fg='blue')} ({time_taken:.1f}s)")
+            print()
+
+
+def exec_steps_parallel(steps: List[Step], workers: int, dag: DAG, strict: Optional[bool] = None) -> None:
+    # put grapher steps in front of the queue to process them as soon as possible and lessen
+    # the load on MySQL
+    steps = sorted(steps, key=lambda x: "grapher://" not in str(x))
+
+    # create execution graph from steps
+    exec_graph = {}
+    steps_str = {str(step) for step in steps}
+    for step in steps:
+        # only add dependencies that are in the list of steps (i.e. are dirty)
+        # NOTE: we have to compare their string versions, the actual objects might have
+        # different attributes
+        exec_graph[str(step)] = {str(dep) for dep in step.dependencies if str(dep) in steps_str}
+
+    exec_graph_parallel(exec_graph, _exec_step_job, workers, dag=dag, strict=strict)
+
+
+def exec_graph_parallel(
+    exec_graph: Dict[str, Any], func: Callable[[str], None], workers: int, use_threads=False, **kwargs
+) -> None:
+    # Create a topological sorter and add tasks with their dependencies
+    topological_sorter = TopologicalSorter(exec_graph)
+    topological_sorter.prepare()
+
+    pool_factory = ThreadPoolExecutor if use_threads else ProcessPoolExecutor
+    with pool_factory(max_workers=workers) as executor:
+        # Dictionary to keep track of future tasks
+        future_to_task: Dict[Future, str] = {}
+
+        while topological_sorter.is_active():
+            # Submit tasks that are ready to the executor
+            for task in topological_sorter.get_ready():
+                future = executor.submit(func, task, **kwargs)
+                future_to_task[future] = task
+
+            # Wait for at least one future to complete
+            done, _ = wait(future_to_task.keys(), return_when=FIRST_COMPLETED)
+
+            # Mark completed tasks as done
+            for future in done:
+                task = future_to_task.pop(future)
+                future.result()
+                topological_sorter.done(task)
+
+
+def _exec_step_job(step_name: str, dag: Optional[DAG] = None, strict: Optional[bool] = None) -> None:
+    """
+    Executes a step.
+
+    :param step_name: The name of the step to execute.
+    :param dag: The original DAG used to create Step object. This must be the same DAG as given to ETL.
+    :param strict: The strictness level for the step execution.
+    """
+    print(f"--- Starting {step_name}", flush=True)
+    assert dag
+    step = parse_step(step_name, dag)
+    strict = _detect_strictness_level(step, strict)
+    with strictness_level(strict):
+        time_taken = timed_run(lambda: step.run())
+
+    print(f"--- Finished {step_name} ({time_taken:.0f}s)", flush=True)
+
+
+def enumerate_steps(steps: List[Step]) -> None:
+    for i, step in enumerate(steps, 1):
+        print(f"{i}. {step}...")
```

Comment:
Personally I have never liked the "..." at the end, cause it looks as if the name of the step is incomplete, and it also makes it less convenient to copy/paste.

Reply:
Makes sense, fixed.